presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java

```diff
@@ -218,60 +221,66 @@ public static PrivilegeGrantInfo toMetastoreApiPrivilegeGrantInfo(HivePrivilegeI
         }
     }
 
-    public static Set<RoleGrant> listApplicableRoles(PrestoPrincipal principal, Function<PrestoPrincipal, Set<RoleGrant>> listRoleGrants)
+    public static Stream<RoleGrant> listApplicableRoles(PrestoPrincipal principal, Function<PrestoPrincipal, Set<RoleGrant>> listRoleGrants)
     {
-        Set<RoleGrant> result = new HashSet<>();
         Queue<PrestoPrincipal> queue = new ArrayDeque<>();
         queue.add(principal);
-        while (!queue.isEmpty()) {
-            PrestoPrincipal current = queue.poll();
-            Set<RoleGrant> grants = listRoleGrants.apply(current);
-            for (RoleGrant grant : grants) {
-                if (!result.contains(grant)) {
-                    result.add(grant);
-                    queue.add(new PrestoPrincipal(ROLE, grant.getRoleName()));
+        Queue<RoleGrant> output = new ArrayDeque<>();
+        Set<RoleGrant> seenRoles = new HashSet<>();
+        return Streams.stream(new AbstractIterator<RoleGrant>() {
```

Comment:
hmm... is the convention to have a new line or not? (diff from https://github.com/prestosql/presto/pull/90/commits/7b73ba0aead016f3b37a1df30b98905a30f4d799)